Flaky Test - BrokerCell created successfully, broker targets config should be updated with broker and triggers

**Describe the bug**
[BrokerCell created successfully, broker targets config should be updated with broker and triggers](https://github.com/google/knative-gcp/blob/571ffebac6611e9b7ca98870a47001434c32cc6b/pkg/reconciler/brokercell/brokercell_test.go#L801) is flaky.

In particular, it asserts the binary data of the ConfigMap must match exactly, but often doesn't. That is because the binary data represents a proto. So while the two protos are logically equivalent, because we are comparing bytes, the serialization order of the maps is asserted. That serialization order is not guaranteed.

 **Expected behavior**
The test should consistently pass.

 **To Reproduce**
```shell
go test ./pkg/reconciler/brokercell/ -count 100
```

 **Additional context**
```
--- FAIL: TestAllCases (0.05s)
    --- FAIL: TestAllCases/BrokerCell_created_successfully,_broker_targets_config_should_be_updated_with_broker_and_triggers (0.01s)
        logger.go:130: 2020-06-23T11:53:39.751-0700     DEBUG   brokercell-controller   reconciler/reconciler.go:171    Creating stats reporter {"events.cloud.google.com/controller": "brokercell-controller"}
        brokercell_test.go:947: Unexpected create (-want, +got):   &v1.ConfigMap{
                TypeMeta:   v1.TypeMeta{},
                ObjectMeta: v1.ObjectMeta{Name: "test-brokercell-brokercell-broker-targets", Namespace: "testnamespace", Labels: map[string]string{"app": "cloud-run-events", "brokerCell": "test-brokercell", "role": "broker-targets"}, OwnerReferences: []v1.OwnerReference{{APIVersion: "internal.events.cloud.google.com/v1alpha1", Kind: "BrokerCell", Name: "test-brokercell", Controller: &true, BlockOwnerDeletion: &true}}},
                Data:       map[string]string{"targets.txt": `brokers:{key:"testnamespace/broker" value:{name:"broker" namespace:"testnamespace" decouple_queue:{topic:"cre-bkr_testnamespace_broker_" subscription:"cre-bkr_testnamespace_broker_"} targets:{key:"trigger1" value:{name:"trigger1" namespace:"testnamespace" broker:"broker" retry_queue:{topic:"cre-tgr_testnamespace_trigger1_" subscription:"cre-tgr_testnamespace_trigger1_"}}} targets:{key:"trigger2" value:{name:"trigger2" namespace:"testnamespace" broker:"broker" retry_queue:{topic:"cre-tgr_testnamespace_trigger2_" subscription:"cre-tgr_testnamespace_trigger2_"}}}}}`},
                BinaryData: map[string][]uint8{
                        "targets": {
                                ... // 94 identical bytes
                                0x74, 0x65, 0x73, 0x74, 0x6e, 0x61, 0x6d, 0x65, 0x73, 0x70, 0x61, 0x63, 0x65, 0x5f, 0x62, 0x72, //  |testnamespace_br|
                                0x6f, 0x6b, 0x65, 0x72, 0x5f, 0x32, 0x71, 0x0a, 0x08, 0x74, 0x72, 0x69, 0x67, 0x67, 0x65, 0x72, //  |oker_2q..trigger|
            -                   0x31,                                                                                           // -|1|
            +                   0x32,                                                                                           // +|2|
                                0x12, 0x65, 0x12, 0x08, 0x74, 0x72, 0x69, 0x67, 0x67, 0x65, 0x72,                               //  |.e..trigger|
            -                   0x31,                                                                                           // -|1|
            +                   0x32,                                                                                           // +|2|
                                0x1a, 0x0d, 0x74, 0x65, 0x73, 0x74, 0x6e, 0x61, 0x6d, 0x65, 0x73, 0x70, 0x61, 0x63, 0x65, 0x22, //  |..testnamespace"|
                                0x06, 0x62, 0x72, 0x6f, 0x6b, 0x65, 0x72, 0x3a, 0x42, 0x0a, 0x1f, 0x63, 0x72, 0x65, 0x2d, 0x74, //  |.broker:B..cre-t|
                                0x67, 0x72, 0x5f, 0x74, 0x65, 0x73, 0x74, 0x6e, 0x61, 0x6d, 0x65, 0x73, 0x70, 0x61, 0x63, 0x65, //  |gr_testnamespace|
                                0x5f, 0x74, 0x72, 0x69, 0x67, 0x67, 0x65, 0x72,                                                 //  |_trigger|
            -                   0x31,                                                                                           // -|1|
            +                   0x32,                                                                                           // +|2|
                                0x5f, 0x12, 0x1f, 0x63, 0x72, 0x65, 0x2d, 0x74, 0x67, 0x72, 0x5f, 0x74, 0x65, 0x73, 0x74, 0x6e, //  |_..cre-tgr_testn|
                                0x61, 0x6d, 0x65, 0x73, 0x70, 0x61, 0x63, 0x65, 0x5f, 0x74, 0x72, 0x69, 0x67, 0x67, 0x65, 0x72, //  |amespace_trigger|
            -                   0x31,                                                                                           // -|1|
            +                   0x32,                                                                                           // +|2|
                                0x5f, 0x32, 0x71, 0x0a, 0x08, 0x74, 0x72, 0x69, 0x67, 0x67, 0x65, 0x72,                         //  |_2q..trigger|
            -                   0x32,                                                                                           // -|2|
            +                   0x31,                                                                                           // +|1|
                                0x12, 0x65, 0x12, 0x08, 0x74, 0x72, 0x69, 0x67, 0x67, 0x65, 0x72,                               //  |.e..trigger|
            -                   0x32,                                                                                           // -|2|
            +                   0x31,                                                                                           // +|1|
                                0x1a, 0x0d, 0x74, 0x65, 0x73, 0x74, 0x6e, 0x61, 0x6d, 0x65, 0x73, 0x70, 0x61, 0x63, 0x65, 0x22, //  |..testnamespace"|
                                0x06, 0x62, 0x72, 0x6f, 0x6b, 0x65, 0x72, 0x3a, 0x42, 0x0a, 0x1f, 0x63, 0x72, 0x65, 0x2d, 0x74, //  |.broker:B..cre-t|
                                0x67, 0x72, 0x5f, 0x74, 0x65, 0x73, 0x74, 0x6e, 0x61, 0x6d, 0x65, 0x73, 0x70, 0x61, 0x63, 0x65, //  |gr_testnamespace|
                                0x5f, 0x74, 0x72, 0x69, 0x67, 0x67, 0x65, 0x72,                                                 //  |_trigger|
            -                   0x32,                                                                                           // -|2|
            +                   0x31,                                                                                           // +|1|
                                0x5f, 0x12, 0x1f, 0x63, 0x72, 0x65, 0x2d, 0x74, 0x67, 0x72, 0x5f, 0x74, 0x65, 0x73, 0x74, 0x6e, //  |_..cre-tgr_testn|
                                0x61, 0x6d, 0x65, 0x73, 0x70, 0x61, 0x63, 0x65, 0x5f, 0x74, 0x72, 0x69, 0x67, 0x67, 0x65, 0x72, //  |amespace_trigger|
            -                   0x32,                                                                                           // -|2|
            +                   0x31,                                                                                           // +|1|
                                0x5f,                                                                                           //  |_|
                        },
                },
              }
{"level":"info","ts":1592938419.803597,"logger":"fallback","caller":"brokercell/reconciler.go:277","msg":"Returned an event","targetMethod":"ReconcileKind","event":"BrokerCell reconciled: \"testnamespace/test-brokercell\""}
{"level":"info","ts":1592938419.8547149,"logger":"fallback","caller":"brokercell/reconciler.go:277","msg":"Returned an event","targetMethod":"ReconcileKind","event":"BrokerCell reconciled: \"testnamespace/test-brokercell\""}
FAIL
FAIL    github.com/google/knative-gcp/pkg/reconciler/brokercell 5.309s
FAIL
```

--- a/pkg/reconciler/brokercell/brokercell_test.go
+++ b/pkg/reconciler/brokercell/brokercell_test.go
@@ -948,6 +948,9 @@ func TestBrokerTargetsReconcileConfig(t *testing.T) {
 		NewTrigger("trigger1", testNS, "broker", WithTriggerSetDefaults),
 		NewTrigger("trigger2", testNS, "broker", WithTriggerSetDefaults))
 	gotMap, err := client.CoreV1().ConfigMaps(testNS).Get(resources.Name(bc.Name, targetsCMName),metav1.GetOptions{})
+	if err != nil {
+		t.Fatalf("Failed to get ConfigMap from client: %v", err)
+	}
 	// compare the ObjectMeta field
 	if diff := cmp.Diff(wantMap.ObjectMeta, gotMap.ObjectMeta); diff != "" {
 		t.Fatalf("Unexpected ObjectMeta in ConfigMap(-want, +got): %s", diff)
@@ -956,10 +959,10 @@ func TestBrokerTargetsReconcileConfig(t *testing.T) {
 	var wantBrokerTargets config.TargetsConfig
 	var gotBrokerTargets config.TargetsConfig
 	if err := proto.Unmarshal(wantMap.BinaryData[targetsCMKey], &wantBrokerTargets); err != nil {
-		t.Fatalf("Failed to deserialize the binary data in configMap: %v", err)
+		t.Fatalf("Failed to deserialize the binary data in ConfigMap: %v", err)
 	}
 	if err := proto.Unmarshal(gotMap.BinaryData[targetsCMKey], &gotBrokerTargets); err != nil {
-		t.Fatalf("Failed to deserialize the binary data in configMap: %v", err)
+		t.Fatalf("Failed to deserialize the binary data in ConfigMap: %v", err)
 	}
 	// compare the broker targets config
 	if diff := cmp.Diff(wantBrokerTargets.String(), gotBrokerTargets.String()); diff != "" {

--- a/pkg/reconciler/brokercell/brokercell_test.go
+++ b/pkg/reconciler/brokercell/brokercell_test.go
@@ -24,25 +24,33 @@ import (
 	appsv1 "k8s.io/api/apps/v1"
 	hpav2beta2 "k8s.io/api/autoscaling/v2beta2"
 	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/runtime"
 	"k8s.io/client-go/kubernetes/scheme"
 	clientgotesting "k8s.io/client-go/testing"
 
+	fakekubeclient "knative.dev/pkg/client/injection/kube/client/fake"
 	"knative.dev/pkg/configmap"
 	"knative.dev/pkg/controller"
 	logtesting "knative.dev/pkg/logging/testing"
 	. "knative.dev/pkg/reconciler/testing"
 
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
 	intv1alpha1 "github.com/google/knative-gcp/pkg/apis/intevents/v1alpha1"
+	"github.com/google/knative-gcp/pkg/broker/config"
 	bcreconciler "github.com/google/knative-gcp/pkg/client/injection/reconciler/intevents/v1alpha1/brokercell"
 	"github.com/google/knative-gcp/pkg/reconciler"
 	"github.com/google/knative-gcp/pkg/reconciler/brokercell/testingdata"
 	. "github.com/google/knative-gcp/pkg/reconciler/testing"
+
+	"google.golang.org/protobuf/proto"
 )
 
 const (
 	testNS         = "testnamespace"
 	brokerCellName = "test-brokercell"
+	targetsCMKey  = "targets"
 )
 
 var (
@@ -798,41 +806,6 @@ func TestAllCases(t *testing.T) {
 			},
 		},
 		{
-			Name: "BrokerCell created successfully, broker targets config should be updated with broker and triggers",
-			Key:  testKey,
-			Objects: []runtime.Object{
-				NewBrokerCell(brokerCellName, testNS, WithBrokerCellSetDefaults),
-				NewBroker("broker", testNS, WithBrokerSetDefaults),
-				NewTrigger("trigger1", testNS, "broker", WithTriggerSetDefaults),
-				NewTrigger("trigger2", testNS, "broker", WithTriggerSetDefaults),
-				NewEndpoints(brokerCellName+"-brokercell-ingress", testNS,
-					WithEndpointsAddresses(corev1.EndpointAddress{IP: "127.0.0.1"})),
-				testingdata.IngressDeploymentWithStatus(t),
-				testingdata.IngressServiceWithStatus(t),
-				testingdata.FanoutDeploymentWithStatus(t),
-				testingdata.RetryDeploymentWithStatus(t),
-				testingdata.IngressHPA(t),
-				testingdata.FanoutHPA(t),
-				testingdata.RetryHPA(t),
-			},
-			WantCreates: []runtime.Object{
-				testingdata.Config(t, NewBrokerCell(brokerCellName, testNS, WithBrokerCellSetDefaults),
-					NewBroker("broker", testNS, WithBrokerSetDefaults),
-					NewTrigger("trigger1", testNS, "broker", WithTriggerSetDefaults),
-					NewTrigger("trigger2", testNS, "broker", WithTriggerSetDefaults)),
-			},
-			WantStatusUpdates: []clientgotesting.UpdateActionImpl{
-				{Object: NewBrokerCell(brokerCellName, testNS,
-					WithBrokerCellReady,
-					WithIngressTemplate("http://test-brokercell-brokercell-ingress.testnamespace.svc.cluster.local/{namespace}/{name}"),
-				)},
-			},
-			WantEvents: []string{
-				configmapCreatedEvent,
-				brokerCellReconciledEvent,
-			},
-		},
-		{
 			Name: "googlecloud created BrokerCell shouldn't be gc'ed because there are brokers",
 			Key:  testKey,
 			Objects: []runtime.Object{
@@ -934,4 +907,65 @@ func TestAllCases(t *testing.T) {
 func emptyHPASpec(template *hpav2beta2.HorizontalPodAutoscaler) *hpav2beta2.HorizontalPodAutoscaler {
 	template.Spec = hpav2beta2.HorizontalPodAutoscalerSpec{}
 	return template
+}
+
+// The unit test to test when the brokerCell created successfully, the broker targets config should be updated with broker
+// and trigger. Since the serialization order of the binary data of brokerTargets in the configMap is not guaranteed, we need
+// to deserialization the binary data to a brokerTargets proto to compare, so it should be rewritten without using the tableTest Utility.
+func TestBrokerTargetsReconcileConfig(t *testing.T) {
+	setReconcilerEnv()
+	bc := NewBrokerCell(brokerCellName, testNS, WithBrokerCellSetDefaults)
+	objects:= []runtime.Object{
+		bc,
+		NewBroker("broker", testNS, WithBrokerSetDefaults),
+		NewTrigger("trigger1", testNS, "broker", WithTriggerSetDefaults),
+		NewTrigger("trigger2", testNS, "broker", WithTriggerSetDefaults),
+	}
+	ctx, _ := SetupFakeContext(t)
+	cmw := configmap.NewStaticWatcher()
+	ctx, client := fakekubeclient.With(ctx,)
+	base := reconciler.NewBase(ctx, controllerAgentName, cmw)
+	testingListers := NewListers(objects)
+	ls := listers{
+		brokerLister:     testingListers.GetBrokerLister(),
+		hpaLister:        testingListers.GetHPALister(),
+		triggerLister:    testingListers.GetTriggerLister(),
+		configMapLister:  testingListers.GetConfigMapLister(),
+		serviceLister:    testingListers.GetK8sServiceLister(),
+		endpointsLister:  testingListers.GetEndpointsLister(),
+		deploymentLister: testingListers.GetDeploymentLister(),
+		podLister:        testingListers.GetPodLister(),
+	}
+	r, err := NewReconciler(base, ls)
+	if err != nil {
+		t.Fatalf("Failed to created BrokerCell reconciler: %v", err)
+	}
+	// here we only want to test the functionality of the reconcileConfig that it should create a brokerTargets config successfully
+	r.reconcileConfig(ctx, bc)
+	wantMap := testingdata.Config(t, NewBrokerCell(brokerCellName, testNS, WithBrokerCellSetDefaults),
+		NewBroker("broker", testNS, WithBrokerSetDefaults),
+		NewTrigger("trigger1", testNS, "broker", WithTriggerSetDefaults),
+		NewTrigger("trigger2", testNS, "broker", WithTriggerSetDefaults))
+	mapList, err := client.CoreV1().ConfigMaps(testNS).List(metav1.ListOptions{})
+	if err != nil || len(mapList.Items) != 1 {
+		t.Fatalf("Failed to get configMap: %v", err)
+	}
+	gotMap := mapList.Items[0]
+	// compare the ObjectMeta field
+	if diff := cmp.Diff(wantMap.ObjectMeta, gotMap.ObjectMeta, cmpopts.EquateEmpty()); diff != "" {
+		t.Fatalf("Unexpected ObjectMeta in ConfigMap(-want, +got): %s", diff)
+	}
+	// deserialize the binary data to a broker targets config proto
+	var wantBrokerTargets config.TargetsConfig
+	var gotBrokerTargets config.TargetsConfig
+	if err := proto.Unmarshal(wantMap.BinaryData[targetsCMKey], &wantBrokerTargets); err != nil {
+		t.Fatalf("Failed to deserialize the binary data in configMap: %v", err)
+	}
+	if err := proto.Unmarshal(gotMap.BinaryData[targetsCMKey], &gotBrokerTargets); err != nil {
+		t.Fatalf("Failed to deserialize the binary data in configMap: %v", err)
+	}
+	// compare the broker targets config
+	if diff := cmp.Diff(wantBrokerTargets.String(), gotBrokerTargets.String(), cmpopts.EquateEmpty()); diff != "" {
+		t.Fatalf("Unexpected brokerTargets in ConfigMap(-want, +got): %s", diff)
+	}
 }

--- a/pkg/reconciler/brokercell/brokercell_test.go
+++ b/pkg/reconciler/brokercell/brokercell_test.go
@@ -36,11 +36,11 @@ import (
 	. "knative.dev/pkg/reconciler/testing"
 
 	"github.com/google/go-cmp/cmp"
-	"github.com/google/go-cmp/cmp/cmpopts"
 	intv1alpha1 "github.com/google/knative-gcp/pkg/apis/intevents/v1alpha1"
 	"github.com/google/knative-gcp/pkg/broker/config"
 	bcreconciler "github.com/google/knative-gcp/pkg/client/injection/reconciler/intevents/v1alpha1/brokercell"
 	"github.com/google/knative-gcp/pkg/reconciler"
+	"github.com/google/knative-gcp/pkg/reconciler/brokercell/resources"
 	"github.com/google/knative-gcp/pkg/reconciler/brokercell/testingdata"
 	. "github.com/google/knative-gcp/pkg/reconciler/testing"
 
@@ -50,6 +50,7 @@ import (
 const (
 	testNS         = "testnamespace"
 	brokerCellName = "test-brokercell"
+	targetsCMName = "broker-targets"
 	targetsCMKey  = "targets"
 )
 
@@ -938,7 +939,7 @@ func TestBrokerTargetsReconcileConfig(t *testing.T) {
 	}
 	r, err := NewReconciler(base, ls)
 	if err != nil {
-		t.Fatalf("Failed to created BrokerCell reconciler: %v", err)
+		t.Fatalf("Failed to create BrokerCell reconciler: %v", err)
 	}
 	// here we only want to test the functionality of the reconcileConfig that it should create a brokerTargets config successfully
 	r.reconcileConfig(ctx, bc)
@@ -946,13 +947,9 @@ func TestBrokerTargetsReconcileConfig(t *testing.T) {
 		NewBroker("broker", testNS, WithBrokerSetDefaults),
 		NewTrigger("trigger1", testNS, "broker", WithTriggerSetDefaults),
 		NewTrigger("trigger2", testNS, "broker", WithTriggerSetDefaults))
-	mapList, err := client.CoreV1().ConfigMaps(testNS).List(metav1.ListOptions{})
-	if err != nil || len(mapList.Items) != 1 {
-		t.Fatalf("Failed to get configMap: %v", err)
-	}
-	gotMap := mapList.Items[0]
+	gotMap, err := client.CoreV1().ConfigMaps(testNS).Get(resources.Name(bc.Name, targetsCMName),metav1.GetOptions{})
 	// compare the ObjectMeta field
-	if diff := cmp.Diff(wantMap.ObjectMeta, gotMap.ObjectMeta, cmpopts.EquateEmpty()); diff != "" {
+	if diff := cmp.Diff(wantMap.ObjectMeta, gotMap.ObjectMeta); diff != "" {
 		t.Fatalf("Unexpected ObjectMeta in ConfigMap(-want, +got): %s", diff)
 	}
 	// deserialize the binary data to a broker targets config proto
@@ -965,7 +962,7 @@ func TestBrokerTargetsReconcileConfig(t *testing.T) {
 		t.Fatalf("Failed to deserialize the binary data in configMap: %v", err)
 	}
 	// compare the broker targets config
-	if diff := cmp.Diff(wantBrokerTargets.String(), gotBrokerTargets.String(), cmpopts.EquateEmpty()); diff != "" {
+	if diff := cmp.Diff(wantBrokerTargets.String(), gotBrokerTargets.String()); diff != "" {
 		t.Fatalf("Unexpected brokerTargets in ConfigMap(-want, +got): %s", diff)
 	}
 }

